### PR TITLE
fix: [PL-25700]: Escaped Special Characters from the Query

### PR DESCRIPTION
**Pull request by adiyaar24**
## Describe your changes
**Issue** 
Whenever a user entered a special character during the start of the search field , we were getting an error in BE as `Dangling Character at Index 0` , this error was coming from Mongo Itself and that threw 500 in UI .

**Resolution**
To Resolve the above issue , I escaped the special characters from the query that was sent to prevent this behaviour . There is no behaviour change in the search results neither any functionality is impacted.

**Testing**
Tested by adding Users , UG with different names and tried searching them and all of those results came accurately as expected and the same thing I reciprocated in PRE QA to make sure there is no inconsistency and I found none , this was because we already have regex matching available in our Mongo Query by default so things remain the same.

**Changes Introduced In**
- Users 
- UserGroups
- Service Account
- Variable
- Connectors

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52435)

### CODE REVIEW

**Comment by adiyaar24** [comment]
trigger uts

**Comment by adiyaar24** [comment]
trigger ut6

**Comment by adiyaar24** [comment]
trigger ut7

**Comment by adiyaar24** [comment]
trigger ut8

**Comment by adiyaar24** [comment]
trigger ut9

**Comment by adiyaar24** [comment]
trigger ut9

File: 120-ng-manager/modules/ng-variables/src/main/java/io/harness/ng/core/variable/services/impl/VariableServiceImpl.java [hunk 1]
```diff
@@ -131,9 +132,10 @@ private Criteria getCriteriaForVariableList(String accountIdentifier, String org
     }
     if (!StringUtils.isEmpty(searchTerm)) {
       criteria = criteria.orOperator(
-          Criteria.where(VariableKeys.name).regex(searchTerm, NGResourceFilterConstants.CASE_INSENSITIVE_MONGO_OPTIONS),
+          Criteria.where(VariableKeys.name)
+              .regex(Pattern.quote(searchTerm), NGResourceFilterConstants.CASE_INSENSITIVE_MONGO_OPTIONS),
           Criteria.where(VariableKeys.identifier)
-              .regex(searchTerm, NGResourceFilterConstants.CASE_INSENSITIVE_MONGO_OPTIONS));
+              .regex(Pattern.quote(searchTerm), NGResourceFilterConstants.CASE_INSENSITIVE_MONGO_OPTIONS));
```
Comment (meenakshi-harness) [inline 120-ng-manager/modules/ng-variables/src/main/java/io/harness/ng/core/variable/services/impl/VariableServiceImpl.java:138]: Please add UTs for cases we are solving with this PR.

File: 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java [hunk 2]
```diff
@@ -1013,12 +1014,12 @@ private PageResponse<Invite> getInvitePage(String accountIdentifier, String orgI
...
     if (!isBlank(searchTerm)) {
-      Page<UserInfo> userInfos = ngUserService.listCurrentGenUsers(
-          accountIdentifier, searchTerm, org.springframework.data.domain.PageRequest.of(0, DEFAULT_PAGE_SIZE));
+      Page<UserInfo> userInfos = ngUserService.listCurrentGenUsers(accountIdentifier, Pattern.quote(searchTerm),
+          org.springframework.data.domain.PageRequest.of(0, DEFAULT_PAGE_SIZE));
       List<String> emailIds = userInfos.stream().map(UserInfo::getEmail).collect(toList());
       Criteria searchTermCriteria = new Criteria();
-      searchTermCriteria.orOperator(
-          Criteria.where(InviteKeys.email).regex(searchTerm), Criteria.where(InviteKeys.email).in(emailIds));
+      searchTermCriteria.orOperator(Criteria.where(InviteKeys.email).regex(Pattern.quote(searchTerm)),
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java:1021]: Same for all services

**Comment by adiyaar24** [comment]
trigger uts

File: 120-ng-manager/modules/ng-variables/src/test/java/io/harness/ng/core/variable/services/impl/VariableServiceImplTest.java [hunk 3]
```diff
@@ -546,4 +547,68 @@ public void testDeleteBatch() {
...
+    List<Variable> varList = new ArrayList<>();
+    varList.add(varA);
+    VariableResponseDTO variableResponseDTO = getVariableResponseDTO(varID, null, null, varID);
+    when(variableMapper.toResponseWrapper(varA)).thenReturn(variableResponseDTO);
+    when(variableRepository.findAll(any(), any())).thenReturn(getPage(varList, 1));
+    PageResponse<VariableResponseDTO> list =
+        variableService.list(accountIdentifier, null, null, specialCharSearchTerm, false, pageable);
+    verify(variableRepository, times(1)).findAll(any(), any());
+    assertThat(list.getContent().size()).isEqualTo(varList.size());
+    assertThat(list.getContent().get(0).getVariable().getIdentifier()).isEqualTo(varID);
```
Comment (meenakshi-harness) [inline 120-ng-manager/modules/ng-variables/src/test/java/io/harness/ng/core/variable/services/impl/VariableServiceImplTest.java:567]: What are we asserting for specialCharSearchTerm ?

Comment (adiyaar24) [reply 120-ng-manager/modules/ng-variables/src/test/java/io/harness/ng/core/variable/services/impl/VariableServiceImplTest.java:567]: When we used to send search term with special chars , it used to error out giving 500 as response , so now if I try to give that search term in the search field I am asserting that if my response is coming or not and if its coming is it the correct one

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52435)* status: Some of the checks are too old. 7 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/modules/ng-variables/src/test/java/io/harness/ng/core/variable/services/impl/VariableServiceImplTest.java` line 567 at r2](https://reviewable.io/reviews/harness/harness-core/52435#gh-1327345841:-NgIy4eb1CheeTWgPgoY:b-lsxwiq) ([raw file](https://github.com/harness/harness-core/blob/7b20762e5f9d909af0242d6c375ab55e35351435/120-ng-manager/modules/ng-variables/src/test/java/io/harness/ng/core/variable/services/impl/VariableServiceImplTest.java#L567)):*
<details><summary><i>Previously, adiyaar24 (Aditya Kashyap) wrote…</i></summary><blockquote>

When we used to send search term with special chars , it used to error out giving 500 as response , so now if I try to give that search term in the search field I am asserting that if my response is coming or not and if its coming is it the correct one 
</blockquote></details>

findAll() call is mocked, which is the call to DB. If this is what was causing 500 error then this test won't help much since the original call is mocked. What you can check in UT is that string used to form query is wrapped by \Q and \E
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 1017 at r2](https://reviewable.io/reviews/harness/harness-core/52435#-NgIzpeXBe5DosoaCeba:-NgIzpeXBe5DosoaCebb:b-odsbqm) ([raw file](https://github.com/harness/harness-core/blob/7b20762e5f9d909af0242d6c375ab55e35351435/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L1017)):*
> ```Java
>     }
>     if (!isBlank(searchTerm)) {
>       Page<UserInfo> userInfos = ngUserService.listCurrentGenUsers(accountIdentifier, Pattern.quote(searchTerm),
> ```

We should use this only when forming criteria.

**Comment by bot2-harness** [comment]
This PR has been automatically closed due to inactivity.

**Pull request closed without merging**